### ISSUE TITLE
Why CommandLineFlag dtor is not virtual?

### ISSUE THREAD

**Issue by MBkkt** [abseil/abseil-cpp#1111]
https://github.com/abseil/abseil-cpp/blob/master/absl/flags/commandlineflag.h#L156

**Comment by derekmauro** [comment]
Good question.

First I would refer you to the flag documentation and tips we've published about using flags:
https://abseil.io/docs/cpp/guides/flags
https://abseil.io/tips/45
https://abseil.io/tips/103

If you read the tips, you'll notice we try to discourage the use of flags in many situations. Flags are globals, and come with all the typical pitfalls of using global state.

The design of flags as globals dates back to the early 2000s. If we were to design a flag library today, I doubt we would pick a design like this, but the design is so pervasive in Google programs that we are stuck with it for now.

The answer your question about why the CommandLineFlag destructor isn't virtual is because we have to get away with not destructing it. If you think about the design of a class meant to be used as a global variable, first it should be constant initialized (`constexpr`) so that it is always safe to access (no [static initialization order fiasco](https://en.cppreference.com/w/cpp/language/siof)), despite the fact that its value before parsing flags in `main()` almost certainly isn't what you want. Marking it `constexpr` precludes having a virtual destructor (at least in C++11). We also very likely have programs accessing flags during the global destructor shutdown sequence, and those programs would fail/crash otherwise due to destructor ordering. There is probably more to this answer that could be discussed, but hopefully this is the gist of it.

In short, this is not a design I would recommend copying, and I'm not even 100% sure how well-defined our behavior is here, but it is working in practice.

**Comment by MBkkt** [comment]
Thanks for such a detailed answer!
I had this question rather than because I decided to use flags in new code, but because some code that builds with Werror cannot build, because of this dtor
